feat!: Introduce the new data dependency loading systems

kristofbolyai opened the pull request:
This branch collects all PRs related to the refactors in data loading, which are aimed to improve dtra loading both systematically and for the user with new UIs.

kristofbolyai committed:
feat: Introduce graphed, dependency based downloads (#2797)

`DownloadManager` has the documentation I'd put in the PR description.

TLDR; The goal is to eliminate download dependency race once and for all, all while taking benefit of having a manager behind all downloads, which include proper failure handling / retries, maximum parallel download settings, a clear and easily calculated download progress percentage and a list of download failures which can be shown to the user.

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

kristofbolyai committed:
Merge branch 'main' into data-loading

kristofbolyai commented:
@ShadowCat117 would you be interested in creating some user-facing system for this PR? Either a download progress notification at the top of the screen, like resource packs, some kind of UI, or even some commands, alternatively. I do think that "click to reload data" is possible with this PR, so maybe we could add an UI for that. However, it might never be used, so maybe just a notification and a "not all of your data downloaded correctly" popup is enough.

ShadowCat117 commented:
Yeah sure I can look into something

magicus commented:
Or maybe just a semi-transparent spinner while we are still downloading data? A progress bar might be tricky since we'd basically need to know how many bytes to download. Downloading the map tiles will take long time, and downloading e.g. label json will be quick.

kristofbolyai committed:
feat: Handle hash conflicts in url files (#2852)

This PR aims to handle hash conflicts in the 3 versions of our URL list in a way that we always aim to re-download the data, if we are not "sure" our hash is correct. During the PR, I also discovered that loading data sources in constructors are VERY race-prone, as it's very likely the constructors run before the remote urls finished downloading, essentially limiting the mod to only use the local cache/bundled urls.json file.

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

kristofbolyai committed:
fix: fix UrlManager crashing with forceSource not being set

kristofbolyai committed:
fix: Fix race condition in DownloadManager (#2865)

ShadowCat117 committed:
fix: Increase delay between territory updates when not in a guild (#2857)

kristofbolyai committed:
feat: Add SystemPropertiesManager to handle custom manager JVM properties (#2861)

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

ShadowCat117 committed:
feat: Front end for download system (#2870)

Co-authored-by: Kristof <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
Merge branch 'main' into data-loading

kristofbolyai committed:
fix: Simplify TerritoryModel no-guild loading delay (#2875)
